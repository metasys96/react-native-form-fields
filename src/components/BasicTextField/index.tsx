import React, {useState} from 'react';
import {View, TextInput, Text, Platform, StyleSheet,ViewStyle,StyleProp,TextStyle} from 'react-native';


interface AppProps {
  label?: string;
  labelColor?: string;
  labelStyle?:  StyleProp<TextStyle>;
  isRequired?: boolean;
  value?: string;
  isError?: boolean;
  onFocusBorderColor?: string;
  errorText?: string;
  errorColor?: string;
  errorTextStyle?:  StyleProp<TextStyle>;
  errorContainerStyle?: StyleProp<ViewStyle>;
  hintText?: string;
  hintTextStyle?: StyleProp<TextStyle>;
  isSecureEntry?: boolean;
  placeholder?: string;
  keyboardType?: string;
  maxLength?: number;
  isClearTextOnFocus?: boolean;
  onChange?: (value: string)=> void;
  textFieldStyle?: StyleProp<ViewStyle>;
  textFieldBorderColor?: string;
  textFieldContainerStyle?: StyleProp<ViewStyle>;
  asteriskTextStyle?: StyleProp<TextStyle>;
  headerStyle?: StyleProp<ViewStyle>;
  textInputProps?: any;
}

const BasicTextField = (props: AppProps) => {
  const {
    label,
    labelColor,
    labelStyle,
    isRequired,
    value,
    isError,
    onFocusBorderColor,
    errorText,
    errorColor,
    errorTextStyle,
    errorContainerStyle,
    hintText,
    hintTextStyle,
    isSecureEntry,
    placeholder,
    keyboardType,
    maxLength,
    isClearTextOnFocus,
    onChange,
    textFieldStyle,
    textFieldBorderColor,
    textFieldContainerStyle,
    asteriskTextStyle,
    headerStyle,
    textInputProps,
  } = props;

  const [isHint, setIsHint] = useState(false);
  const [outerBorderColor, setOuterBorderColor] = useState(
    textFieldBorderColor,
  );
  const [textValue, setTextValue] = useState(value);

  const onChangeTextInput = (text: string) => {
    setTextValue(text);

    if ('function' === typeof onChange) {
      onChange(text);
    }
  };

  const onFocus = () => {
    setIsHint(true);
    setOuterBorderColor(onFocusBorderColor);
  };
  const onBlur = () => {
    setIsHint(false);
    setOuterBorderColor(textFieldBorderColor);
  };

  return (
    <View>
      <View style={StyleSheet.flatten([DefaultStyles.headerContainerStyle, headerStyle])}>
        <View style={{justifyContent: 'center'}}>
          <Text
            style={StyleSheet.flatten([
              {color: isError ? errorColor : labelColor, fontSize: 16},
              labelStyle,
            ])}>
            {label}
          </Text>
        </View>

        <View style={{justifyContent: 'center'}}>
          {isRequired ? (
            <Text
              style={StyleSheet.flatten([
                {color: 'red', fontSize: 16},
                asteriskTextStyle,
              ])}>
              *
            </Text>
          ) : null}
        </View>

        {isHint ? (
          <View style={{justifyContent: 'center'}}>
            <Text
              style={StyleSheet.flatten([
                DefaultStyles.hintTextStyle,
                hintTextStyle,
              ])}>
              {hintText}
            </Text>
          </View>
        ) : null}
      </View>

      <View
        style={[
          {
            borderColor: isError ? errorColor : outerBorderColor,
            borderWidth: 1,
            borderRadius: 10,
          },
          textFieldContainerStyle,
        ]}>
        <TextInput
          maxLength={maxLength}
          secureTextEntry={isSecureEntry}
          keyboardType={keyboardType}
          style={StyleSheet.flatten([DefaultStyles.textFieldDefaultStyle, textFieldStyle])}
          placeholder={placeholder}
          onFocus={onFocus}
          onBlur={onBlur}
          value={textValue}
          onChangeText={onChangeTextInput}
          clearTextOnFocus={isClearTextOnFocus}
          {...textInputProps}
        />
      </View>

      {isError ? (
        <View
          style={StyleSheet.flatten([
            {justifyContent: 'center', marginLeft: '1%'},
            errorContainerStyle,
          ])}>
          <Text
            style={StyleSheet.flatten([errorTextStyle, {color: errorColor}])}>
            {errorText}
          </Text>
        </View>
      ) : (
        <Text> </Text>
      )}
    </View>
  );
};

const DefaultStyles = StyleSheet.create({
  hintTextStyle:{
    color: 'orange', 
    fontSize: 16, 
    marginLeft: '2%'
  },
  textFieldDefaultStyle : {
    color: 'black',
    paddingLeft: '2%',
    height: 40,
    fontSize: 20,
    paddingVertical: Platform.OS == 'android' ? 0 : null,
  },
  headerContainerStyle: {
    width: '100%',
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginVertical: '0.5%',
    paddingLeft: '1%',
  }
})

BasicTextField.defaultProps = {
  textFieldBorderColor: 'black',
  label: '',
  isRequired: false,
  value: '',
  isError: false,
  errorColor: 'red',
  onFocusBorderColor: '#007AFE',
  errorText: '',
  hintText: '',
  isSecureEntry: false,
  placeholder: '',
  keyboardType: 'default',
  maxLength: null,
  isClearTextOnFocus: false,
  labelColor: 'black',
};

export default BasicTextField;
