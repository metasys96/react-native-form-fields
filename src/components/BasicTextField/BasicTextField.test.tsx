import {shallow} from 'enzyme';
import BasicTextField from './index';
import React from 'react';
import 'react-native';
import {cleanup, fireEvent} from 'react-native-testing-library';
import renderer from 'react-test-renderer';

afterEach(cleanup);

function getShallowComponent() {
  return shallow(<BasicTextField isSecureEntry={true} value="aditya" />);
}

function getJSONRenderer() {
  return renderer.create(<BasicTextField />).toJSON();
}

test('create snapshot', () => {
  const component = getJSONRenderer();
  expect(component).toMatchSnapshot();
});

it('renders component', () => {
  const component = getJSONRenderer();
  expect(component).toBeTruthy();
});

it('should contain one textInput and two texts', () => {
  const component = getShallowComponent();
  expect(component.find('TextInput')).toHaveLength(1);
  expect(component.find('Text').length).toEqual(2);
});

it('should check the property value and secureTextEntry when passed', () => {
  const component = getShallowComponent();
  expect(component.find('TextInput').get(0).props.value).toEqual('aditya');
  expect(component.find('TextInput').get(0).props.secureTextEntry).toBeTruthy();
});

it('should change the value', async () => {
  const onChangeFunction = jest.fn();
  const component = shallow(
    <BasicTextField
      onChange={onChangeFunction}
      isSecureEntry={true}
      value="aditya"
    />,
  );

  const textField = component.find('TextInput').get(0); 
  fireEvent.changeText(textField as any, 'abcdefg');
  expect(onChangeFunction).toHaveBeenCalledTimes(1);
  const textField1 = component.find('TextInput').get(0);
  expect(textField1.props.value).toEqual('abcdefg');
});
