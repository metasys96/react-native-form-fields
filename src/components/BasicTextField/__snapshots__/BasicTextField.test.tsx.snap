
exports[`create snapshot 1`] = `
<View>
  <View
    style={
      Object {
        "alignText": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "marginVertical": "0.5%",
        "paddingLeft": "1%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 16,
          }
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Object {
          "justifyContent": "center",
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "borderColor": "black",
          "borderRadius": 10,
          "borderWidth": 1,
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      clearTextOnFocus={false}
      keyboardType="default"
      maxLength={null}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder=""
      rejectResponderTermination={true}
      secureTextEntry={false}
      style={
        Object {
          "color": "black",
          "fontSize": 20,
          "height": 40,
          "paddingLeft": "2%",
          "paddingVertical": null,
        }
      }
      testID="input"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <Text>
     
  </Text>
</View>
`;
